src/cache.rs

```diff
@@ -0,0 +1,305 @@
+//! This module provides caching utils
+
+use std::{
+  fs::{create_dir_all, remove_dir_all, File, OpenOptions},
+  io::{self, BufWriter, Error, ErrorKind, Seek},
+  ops::Deref,
+  path::PathBuf,
+  sync::{Arc, Mutex},
+};
+
+use abomonation::{encode, Abomonation};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+
+use ff::PrimeField;
+use memmap2::MmapMut;
+use once_cell::sync::Lazy;
+
+use crate::{
+  r1cs::commitment_key_size,
+  supernova::{AuxParams, CircuitDigests, NonUniformCircuit, PublicParams},
+  traits::{
+    circuit_supernova::StepCircuit,
+    commitment::{CommitmentEngineTrait, CommitmentKeyTrait},
+    Group,
+  },
+  CircuitShape, CommitmentKey,
+};
+
+/// Global registry for public parameters and commitment keys that are
+/// cached onto the users local file system
+pub static ARECIBO_CACHE: Lazy<Cache> = Lazy::new(init_cache);
+
+/// If the env var `ARECIBO_CACHE` exists, then use that value.
+/// Otherwise, set `ARECIBO_CACHE` to be `$HOME/.arecibo/`.
+fn init_cache() -> Cache {
+  let path = if let Ok(x) = std::env::var("ARECIBO_CACHE") {
+    tracing::debug!("{:?}", &x);
+    PathBuf::from(x)
+  } else {
+    let path = home::home_dir()
+      .expect("targets without $HOME not supported")
+      .join(".arecibo/");
+    std::env::set_var("ARECIBO_CACHE", &path);
+    path
+  };
+
+  create_dir_all(&path).unwrap();
+  create_dir_all(path.join("commitment_keys")).unwrap();
+  create_dir_all(path.join("public_params")).unwrap();
+
+  Cache {
+    inner: Arc::new(Mutex::new(path)),
+  }
+}
+
+/// Cache structure that holds the root directory of all the cached files
+pub struct Cache {
+  inner: Arc<Mutex<PathBuf>>,
+}
+
+impl Cache {
+  /// Sets the inner root directory of the cache
+  pub fn set_inner(&self, path: &PathBuf) {
+    self.inner.lock().unwrap().clone_from(path);
+
+    create_dir_all(path).unwrap();
+    create_dir_all(path.join("commitment_keys")).unwrap();
+    create_dir_all(path.join("public_params")).unwrap();
+  }
```

Comment:
There's no way to get a cache other than `init_cache()`, which makes me think this should be called `reset_inner`. But as mentioned above, I don't think we need the functionality at this stage.